src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperParseTreeRewriter.cs

```diff
@@ -144,8 +144,11 @@ public override SyntaxNode VisitMarkupElement(MarkupElementSyntax node)
                         // Non-TagHelper tag.
                         ValidateParentAllowsPlainStartTag(startTag);
 
-                        if (!startTag.IsSelfClosing() && !startTag.IsVoidElement())
+                        if (node.EndTag != null || (!startTag.IsSelfClosing() && !startTag.IsVoidElement()))
```

Comment:
So it looks like `(!startTag.IsSelfClosing() && !startTag.IsVoidElement())` is for the case where we see an element that *should* have an end-tag, but it doesn't have one (omitted). I'm trying to find where that tag tracker would get popped and I don't see it.

Reply:
That's right. Consider,
`<p><div>Foo<p></p></p>` where `p` is a tag helper and `div` is not. This condition will be satisfied for `<div>` and will be pushed to the stack but it won't be popped from the stack because there is no end tag. This is necessary because anything that comes after `<div>` will be validated to make sure `div` can be a parent to it. We do that in many places like [this](https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperParseTreeRewriter.cs#L655).